keepcurrent.go

```diff
@@ -40,67 +44,98 @@ type Runner struct {
 
 // New construct a runner which synchronizes data from one source to one or more sinks
 func New(from Source, to ...Sink) *Runner {
-	return &Runner{func(error) {}, func(Sink, error) {}, from, to, time.Time{}}
+	return &Runner{
+		func(error, int) time.Duration { return 0 },
+		func(Sink, error) {},
+		from, to, time.Time{}}
 }
 
 // InitFrom synchronizes data from the given source to configured sinks.
 func (runner *Runner) InitFrom(s Source) {
 	if len(runner.sinks) == 0 {
 		return
 	}
-	runner.syncOnce(s)
+	runner.syncOnce(s, nil)
 }
 
-// Start starts a loop to actually synchronizes data with given interval.  It
+// Start starts the loop to actually synchronizes data with given interval. It
 // returns a function to stop the loop.
 func (runner *Runner) Start(interval time.Duration) func() {
 	if len(runner.sinks) == 0 {
 		return func() {}
 	}
 	tk := time.NewTicker(interval)
-	ch := make(chan struct{})
+	chStop := make(chan struct{})
 	go func() {
 		for {
-			runner.syncOnce(runner.source)
+			runner.syncOnce(runner.source, chStop)
 			select {
-			case <-ch:
+			case <-chStop:
 				tk.Stop()
 				return
 			case <-tk.C:
 			}
 		}
 	}()
-	return func() { close(ch) }
+	return func() { close(chStop) }
 }
 
-func (runner *Runner) syncOnce(from Source) {
-	start := time.Now()
-	rc, err := from.Fetch(runner.lastUpdated)
-	if err == ErrUnmodified {
-		return
-	}
-	if err != nil {
-		runner.OnSourceError(err)
-		return
-	}
-	runner.lastUpdated = start
-	defer rc.Close()
-	if len(runner.sinks) == 1 {
-		s := runner.sinks[0]
-		if err := s.UpdateFrom(rc); err != nil {
-			runner.OnSinkError(s, err)
+func (runner *Runner) syncOnce(from Source, chStop chan struct{}) {
+	for tries := 1; ; tries++ {
+		start := time.Now()
+		rc, err := from.Fetch(runner.lastUpdated)
+		if err == ErrUnmodified {
+			return
+		}
+		var buf bytes.Buffer
+		if err == nil {
+			// Read ahead to surface any error reading from the source
+			_, err = io.Copy(&buf, rc)
```

Comment:
I wonder if this can fix https://github.com/getlantern/lantern-internal/issues/3593#issuecomment-617167114. Previously, any error downloading the file is not treated as source error.

Reply:
Yeah makes sense, though it would make chaining sources or sinks harder. For example, `tarGzSource` reads the archive incrementally and returns once it finds the file expected, without requiring the whole archive being available.

Reply:
I should admit that part of the reason is that the `archiver` package https://github.com/mholt/archiver/blob/master/targz.go doesn't have an API to accept byte slice.

Reply:
Ah, good point about chaining, I didn't consider that.  An alternative could be to add a flag to the Runner to `SkipPreRead` (or something with a better name =P).  But that could just be added later if some user of the package desires.